toolkit/tools/internal/retry/retry.go

```diff
@@ -40,21 +41,25 @@ func calculateLinearDelay(failCount int, sleep time.Duration) time.Duration {
 	return sleep * time.Duration(failCount)
 }
 
-// backoffSleep sleeps for the given duration, unless the cancel channel is triggered or closed first. The cancel channel
-// may be nil in which case the sleep will always complete.
-func backoffSleep(delay time.Duration, cancel <-chan struct{}) (cancelled bool) {
+// backoffSleep sleeps for the given duration, unless the context is cancelled. The context
+// may be nil in which case the background context will be used and the sleep will always complete.
+func backoffSleep(delay time.Duration, ctx context.Context) (cancelled bool) {
+	// Context is optional, if not provided, use a background context.
+	if ctx == nil {
```

Comment:
ended up moving this back into the private function since there are multiple paths that reach here from public functions.